### PR TITLE
Fix #722: Introduced Walkthrough

### LINKED ISSUE #722: Lowfi Walkthrough - Part 1
Mock: https://xd.adobe.com/view/0687f00c-695b-437a-79a6-688e7f4f7552-70b6/screen/9bb5c0e5-f77e-4c28-8d79-18515d0e9711/WT-Walkthrough-1-3/

At first we need only following things:
- [ ] WalkthroughActivity
- [ ] WalkthroughActivityPresenter
- [ ] WalkthroughFragment
- [ ] WalkthroughFragmentPresenter

The walkthrough fragment will contain following things
- [ ] Header - This should be functional
Screenshot 2020-02-27 at 2 35 21 PM
- [ ] Dummy button on click of which the header should finish 1/3 progress.

Don't include anything related to content right now.

Also, add test cases too.

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Add Walkthrough Base Classes

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by aggarwalpulkit596** [comment]
@rt4914 before i move forward with anything else can you check whether everything is fine or not.I thought this way it would be easier for you to review the PR as well as for me make the changes

**Comment by rt4914** [comment]
> @rt4914 before i move forward with anything else can you check whether everything is fine or not.I thought this way it would be easier for you to review the PR as well as for me make the changes

@aggarwalpulkit596 This is very awesome, you have laid out all the files perfectly. Though there are some nit changes but for now I am just focusing on the approach, which is correct. Thanks a lot.

**Comment by aggarwalpulkit596** [comment]
@rt4914 just one more thing is there any custom progress bar like in the mock because I saw that progress bar in some more mocks but couldn't find it anywhere in the project

**Comment by rt4914** [comment]
> @rt4914 just one more thing is there any custom progress bar like in the mock because I saw that progress bar in some more mocks but couldn't find it anywhere in the project

The one that you will be working on, there is nothing similar to that, you might have seen something like this in AudioPlayer but actually that is a seekbar not a progress bar. You will need to create your own progress bar for this.

**Comment by aggarwalpulkit596** [comment]
@rt4914 i've completed the work can you tell about the test cases what kind of test cases should i write and for which components should i write them

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -13,6 +13,8 @@
     android:roundIcon="@mipmap/ic_launcher_round"
     android:supportsRtl="true"
     android:theme="@style/OppiaTheme">
+    <activity android:name=".walkthrough.WalkthroughActivity"
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: This should come in alphabetical order. Mostly at the end of all other activities.

File: app/src/main/java/org/oppia/app/activity/ActivityComponent.kt [hunk 2]
```diff
@@ -66,6 +67,7 @@ interface ActivityComponent {
   fun inject(explorationTestActivity: ExplorationTestActivity)
   fun inject(helpActivity: HelpActivity)
   fun inject(homeActivity: HomeActivity)
+  fun inject(walkthroughActivity: WalkthroughActivity)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/activity/ActivityComponent.kt]: This should also follow alphabetical order.

File: app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt [hunk 3]
```diff
@@ -58,6 +59,7 @@ interface FragmentComponent {
   fun inject(myDownloadsFragment: MyDownloadsFragment)
   fun inject(navigationDrawerFragment: NavigationDrawerFragment)
   fun inject(onboardingFragment: OnboardingFragment)
+  fun inject(walkthroughFragment: WalkthroughFragment)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt]: Ditto as above

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivity.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  companion object {
+    fun createWalkthroughActivity(context: Context): Intent {
+      val intent = Intent(context, WalkthroughActivity::class.java)
+      intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+      return intent
+    }
+  }
+
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivity.kt]: Remove 1 extra blank space.

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        R.id.walkthrough_fragment_placeholder,
+        WalkthroughFragment()
+      ).commitNow()
+    }
+  }
+
+  private fun getWalkthroughFragment(): WalkthroughFragment? {
+    return activity.supportFragmentManager.findFragmentById(R.id.walkthrough_fragment_placeholder) as WalkthroughFragment?
+  }
+}
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityPresenter.kt]: EOF is missing, add one blank line at the end.

**Comment by aggarwalpulkit596** [comment]
Screenshot_1582868530

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+  override fun onAttach(context: Context) {
+    super.onAttach(context)
+    fragmentComponent.inject(this)
+  }
+
+  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+    return walkthroughFragmentPresenter.handleCreateView(inflater, container)
+  }
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Remove this blank line.

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import org.oppia.app.fragment.InjectableFragment
+import javax.inject.Inject
+
+/** Fragment that contains an walkthrough of the app. */
+class WalkthroughFragment : InjectableFragment() {
+  @Inject
+  lateinit var walkthroughFragmentPresenter: WalkthroughFragmentPresenter
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Write this in single line:
`@Inject lateinit var walkthroughFragmentPresenter: WalkthroughFragmentPresenter`

**Comment by aggarwalpulkit596** [comment]
@rt4914 for checking progress bar progress we need something like this https://github.com/vanniktech/espresso-utils/blob/master/espresso-core-utils/src/main/java/com/vanniktech/espresso/core/utils/ProgressMatcher.java

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import androidx.appcompat.app.AppCompatActivity
+import androidx.fragment.app.Fragment
+import org.oppia.app.databinding.WalkthroughFragmentBinding
+import org.oppia.app.fragment.FragmentScope
+import javax.inject.Inject
+
+/** The presenter for [WalkthroughFragment]. */
+@FragmentScope
+class WalkthroughFragmentPresenter @Inject constructor(
+  private val activity: AppCompatActivity,
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt]: Remove this unused line

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      it.presenter = this
+    }
+
+    return binding.root
+  }
+
+  fun nextPage() {
+    binding.progressBar.progress = 1
+  }
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt:35]: Remove this blank line

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+}
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt]: EOF is missing, add one blank line at the end.

File: app/src/main/res/drawable/rounded_corners_progress_bar.xml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    android:right="1dp">
+
+    <scale android:scaleWidth="100%">
+      <shape>
+        <corners android:radius="8dp"/>
+        <solid android:color="@color/colorPrimaryDark"/>
+      </shape>
+    </scale>
+  </item>
+</layer-list>
```
Comment (rt4914) [inline app/src/main/res/drawable/rounded_corners_progress_bar.xml]: EOF is missing, add one blank line at the end.

File: app/src/main/res/layout/walkthrough_activity.xml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+  xmlns:tools="http://schemas.android.com/tools"
+  android:id="@+id/walkthrough_fragment_placeholder"
+  android:layout_width="match_parent"
+  android:layout_height="match_parent"
+  tools:context=".walkthrough.WalkthroughActivity" />
```
Comment (rt4914) [inline app/src/main/res/layout/walkthrough_activity.xml]: EOF is missing, add one blank line at the end.

File: app/src/main/res/layout/walkthrough_fragment.xml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+      android:layout_height="wrap_content"
+      android:text="@string/next"
+      android:onClick="@{(v) -> presenter.nextPage()}"
+      style="@style/StateButtonActive"
+      android:layout_marginBottom="32dp"
+      app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintEnd_toEndOf="parent"
+      app:layout_constraintStart_toStartOf="parent" />
+  </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
Comment (rt4914) [inline app/src/main/res/layout/walkthrough_fragment.xml]: EOF is missing, add one blank line at the end.

File: app/src/main/res/layout/walkthrough_fragment.xml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      android:layout_margin="16dp"
+      android:src="@drawable/ic_arrow_back_white_24dp"
+      android:tint="@color/grey"
+      app:layout_constraintStart_toStartOf="parent"
+      app:layout_constraintTop_toTopOf="parent"
+      android:focusable="true"
+      android:contentDescription="@string/back_button" />
+
+    <ProgressBar
+      android:id="@+id/progressBar"
```
Comment (rt4914) [inline app/src/main/res/layout/walkthrough_fragment.xml]: id should be `walkthrough_progress_bar`

File: app/src/main/res/layout/walkthrough_fragment.xml [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+      name="presenter"
+      type="org.oppia.app.walkthrough.WalkthroughFragmentPresenter" />
+  </data>
+
+  <androidx.constraintlayout.widget.ConstraintLayout
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <ImageView
+
```
Comment (rt4914) [inline app/src/main/res/layout/walkthrough_fragment.xml]: Remove this blank line

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+      onView(withId(R.id.progressBar)).check(matches(isDisplayed()))
+    }
+  }
+  @Test
+  fun testWalkthroughFragment_increaseProgress() {
+    ActivityScenario.launch(WalkthroughActivity::class.java).use {
+      onView(withId(R.id.next_button)).perform(click())
+    }
+  }
+}
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: EOF is missing, add one blank line at the end.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@aggarwalpulkit596 Please do not work on code once you have assigned it for review, because it creates issues for the reviewers. If you want to work again, just unassign the reviewer first and then assign it back to you.
Currently I was reviewing your PR and in between you made commits and that brings changes to the code on which currently I am checking, which is not good.

Regarding your implementation, `Next` button is working only once and `Back Image` is not working.

Ideally what we want is that the `Next` should be clickable until the progress is fully finished and the `Back Image` should be clickable until the progress is fully empty. Also, you will need to write test cases regarding that.

**Comment by rt4914** [comment]
@aggarwalpulkit596 Please go through this https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR to understand how `Assignees` work.

**Comment by aggarwalpulkit596** [comment]
> @aggarwalpulkit596 Please do not work on code once you have assigned it for review, because it creates issues for the reviewers. If you want to work again, just unassign the reviewer first and then assign it back to you.
> Currently I was reviewing your PR and in between you made commits and that brings changes to the code on which currently I am checking, which is not good.
> 
> Regarding your implementation, `Next` button is working only once and `Back Image` is not working.
> 
> Ideally what we want is that the `Next` should be clickable until the progress is fully finished and the `Back Image` should be clickable until the progress is fully empty. Also, you will need to write test cases regarding that.

thanks for the explanation wasn't aware about the functionality and will keep that in mind for review code

**Comment by aggarwalpulkit596** [comment]
@rt4914 i've made some changes please go through them once

File: app/src/main/AndroidManifest.xml [hunk 17]
```diff
@@ -119,5 +119,7 @@
     <activity android:name=".testing.ReviewCardFragmentTestActivity"
       android:screenOrientation="portrait"
       android:theme="@style/OppiaThemeWithoutActionBar" />
+    <activity android:name=".walkthrough.WalkthroughActivity"
+      android:screenOrientation="portrait"/>
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Spacing issue: `android:screenOrientation="portrait" />`

Simply use reformatting shortcut:
`Ctrl + Alt + L`

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+  @Test
+  fun testWalkthroughFragment_skipNextButtonIsVisible() {
+    ActivityScenario.launch(WalkthroughActivity::class.java).use {
+      onView(withId(R.id.next_button)).check(matches(isDisplayed()))
+      onView(withId(R.id.walkthrough_progress_bar)).check(matches(isDisplayed()))
+    }
+  }
+  @Test
+  fun testWalkthroughFragment_increaseProgress() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: In this, make sure that you check whether the progress has increased or not.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+import org.junit.runner.RunWith
+import org.oppia.app.R
+import org.oppia.app.onboarding.OnboardingActivity
+
+/** Tests for [WalkthroughFragment]. */
+@RunWith(AndroidJUnit4::class)
+class OnboardingFragmentTest {
+
+  @Before
+  fun setUp() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: You don't need this

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+    Intents.init()
+  }
+
+  @After
+  fun tearDown() {
+    Intents.release()
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: You don't need this.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 16 above)
Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: PTAL

**Review by rt4914 — CHANGES_REQUESTED** [review]
This is very nice implementation. Thanks.

You should write more test cases to correctly verify that on next click the progress bas actually changes and on back click also.

Once this PR is submitted, we will work on second part of this PR.

Thanks.

**Comment by rt4914** [comment]
@aggarwalpulkit596 Always reply to these conversations, if you have solved the issue, simply write `Done.`

Screenshot 2020-02-28 at 1 20 06 PM

**Comment by aggarwalpulkit596** [comment]
> This is very nice implementation. Thanks.
> 
> You should write more test cases to correctly verify that on next click the progress bas actually changes and on back click also.
> 
> Once this PR is submitted, we will work on second part of this PR.
> 
> Thanks.

for checking progress bar progress we'll be needing a custom action similar to the link posted below
https://github.com/vanniktech/espresso-utils/blob/master/espresso-core-utils/src/main/java/com/vanniktech/espresso/core/utils/ProgressMatcher.java

File: app/src/main/res/layout/walkthrough_fragment.xml (hunk 15 above)
Comment (aggarwalpulkit596) [reply app/src/main/res/layout/walkthrough_fragment.xml]: done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityPresenter.kt (hunk 5 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivityPresenter.kt]: done

File: app/src/main/res/layout/walkthrough_activity.xml (hunk 12 above)
Comment (aggarwalpulkit596) [reply app/src/main/res/layout/walkthrough_activity.xml]: done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivity.kt (hunk 4 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughActivity.kt]: done

File: app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt (hunk 3 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt]: done

File: app/src/main/java/org/oppia/app/activity/ActivityComponent.kt (hunk 2 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/activity/ActivityComponent.kt]: done

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (aggarwalpulkit596) [reply app/src/main/AndroidManifest.xml]: done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 7 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt (hunk 8 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragmentPresenter.kt]: done

File: app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 6 above)
Comment (aggarwalpulkit596) [reply app/src/main/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/main/res/drawable/rounded_corners_progress_bar.xml (hunk 11 above)
Comment (aggarwalpulkit596) [reply app/src/main/res/drawable/rounded_corners_progress_bar.xml]: done

File: app/src/main/res/layout/walkthrough_fragment.xml (hunk 14 above)
Comment (aggarwalpulkit596) [reply app/src/main/res/layout/walkthrough_fragment.xml]: done

**Comment by aggarwalpulkit596** [comment]
@rt4914 can you help out in writing test for progress bar all I could find was this 
https://stackoverflow.com/questions/57531678/assert-progress-of-progressbar-in-espresso-test
 but this doesn't look good

**Comment by rt4914** [comment]
@aggarwalpulkit596 Everything you need regarding the progress test case is mentioned in the link shared by you: https://github.com/vanniktech/espresso-utils/blob/master/espresso-core-utils/src/main/java/com/vanniktech/espresso/core/utils/ProgressMatcher.java

Just introduce `ProgressMatcher` in `utility` package of test cases and try using that in your test case, which is mentioned [here](https://github.com/vanniktech/espresso-utils/blob/1d1f6424cc2eaf8233a901c48f17e1c99626e8a0/espresso-core-utils/src/main/java/com/vanniktech/espresso/core/utils/ProgressMatcher.java#L14).

**Comment by rt4914** [comment]
@aggarwalpulkit596 Have a look at it: https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section and make sure that the `Assignees` section points to correct person.

Also, you have haven't replied on the conversations yet. Please check that and reply accordingly.

**Comment by aggarwalpulkit596** [comment]
> @aggarwalpulkit596 Have a look at it: https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section and make sure that the `Assignees` section points to correct person.
> 
> Also, you have haven't replied on the conversations yet. Please check that and reply accordingly.

Instead of Replying on each of issues can i just resolve the conversation ?

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 20 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Got it

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 19 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 18 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/main/AndroidManifest.xml (hunk 17 above)
Comment (aggarwalpulkit596) [reply app/src/main/AndroidManifest.xml]: done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 16 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/sharedTest/java/org/oppia/app/utility/ProgressMatcher.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
+package org.oppia.app.utility;
+
+import androidx.annotation.CheckResult;
+import androidx.test.espresso.matcher.BoundedMatcher;
+import android.view.View;
+import android.widget.ProgressBar;
+import org.hamcrest.Description;
+
+
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/utility/ProgressMatcher.java]: Convert this file to Kotlin. Remove this blank line above. Also, as mentioned earlier in one of the files, always write Documentation for new classes. Also, In this particular case please mention the reference URL also.

Check `DrawableMatcher` for reference.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+  @Test
+  fun testWalkthroughFragment_increaseProgress() {
+    ActivityScenario.launch(WalkthroughActivity::class.java).use {
+      onView(withId(R.id.next_button)).perform(click())
+      onView(withId(R.id.walkthrough_progress_bar)).check(matches(withProgress(1)))
+    }
+  }
+
+  @Test
+  fun testWalkthroughFragment_increaseAndDecreaseProgress() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Rename this to `testWalkthroughFragment_increaseProgress_decreaseProgress_progressWorksCorrectly`

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+  @Test
+  fun testWalkthroughFragment_skipNextButtonIsVisible() {
+    ActivityScenario.launch(WalkthroughActivity::class.java).use {
+      onView(withId(R.id.next_button)).check(matches(isDisplayed()))
+      onView(withId(R.id.walkthrough_progress_bar)).check(matches(isDisplayed()))
+    }
+  }
+  @Test
+  fun testWalkthroughFragment_increaseProgress() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Rename: `testWalkthroughFragment_increaseProgress _worksCorrectly`

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+import org.junit.runner.RunWith
+import org.oppia.app.R
+import org.oppia.app.utility.ProgressMatcher.withProgress
+
+/** Tests for [WalkthroughFragment]. */
+@RunWith(AndroidJUnit4::class)
+class OnboardingFragmentTest {
+
+  @Test
+  fun testWalkthroughFragment_skipNextButtonIsVisible() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: testWalkthroughFragment_checkSkipButton_skipNextButtonIsVisible

Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Normal convention for any test case:
_`testAction_withOneCondition_withSecondCondition_hasExpectedOutcome`_

**Review by rt4914 — CHANGES_REQUESTED** [review]
Looks good, some changes suggested.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 24 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 23 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 22 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/sharedTest/java/org/oppia/app/utility/ProgressMatcher.java (hunk 21 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/utility/ProgressMatcher.java]: done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Test
+  fun testWalkthroughFragment_checkSkipButton_skipNextButton_progressBar_IsVisible() {
+    ActivityScenario.launch(WalkthroughActivity::class.java).use {
+      onView(withId(R.id.next_button)).check(matches(isDisplayed()))
+      onView(withId(R.id.back_button)).check(matches(isDisplayed()))
+      onView(withId(R.id.walkthrough_progress_bar)).check(matches(isDisplayed()))
+    }
+  }
+  @Test
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt:28]: Add one extra blank line above this.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt [hunk 26]
```diff
@@ -0,0 +1,43 @@
+package org.oppia.app.walkthrough
+
+import androidx.test.core.app.ActivityScenario
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.*
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: Avoid using * in imports.

Better use these imports:
```
import androidx.test.core.app.ActivityScenario
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Test
import org.junit.runner.RunWith
import org.oppia.app.R
import org.oppia.app.utility.ProgressMatcher.Companion.withProgress
```

**Review by rt4914 — CHANGES_REQUESTED** [review]
I think this should be the final change. Please focus on formatting related issues.

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 26 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt]: done

File: app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt (hunk 25 above)
Comment (aggarwalpulkit596) [reply app/src/sharedTest/java/org/oppia/app/walkthrough/WalkthroughFragment.kt:28]: done

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Comment by rt4914** [comment]
@aggarwalpulkit596 Once it is ready for merge, I will merge it and then you can work on #729 . Take fresh copy of develop once this PR is merged, create a new branch and then work on #729

**Review by rt4914 — APPROVED** [review]
LGTM everything is correct, but somehow test cases are failing on CircleCi. Will need to have a look at it.

**Comment by aggarwalpulkit596** [comment]
@rt4914 test is failing for Splash screen and I haven't made any change to that

**Comment by aggarwalpulkit596** [comment]
I guess there some issue with the Emulator due to which some test becomes unresponsive and hence fails

**Comment by aggarwalpulkit596** [comment]
@rt4914 Can you take a look t this because I can't start the next issue as it depends upon this PR.

File: app/src/main/AndroidManifest.xml [hunk 27]
```diff
@@ -119,5 +119,7 @@
     <activity android:name=".testing.ReviewCardFragmentTestActivity"
       android:screenOrientation="portrait"
       android:theme="@style/OppiaThemeWithoutActionBar" />
+    <activity android:name=".walkthrough.WalkthroughActivity"
+      android:screenOrientation="portrait" />
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml:123]: Just press
`Control + Alt + L` to reformat everything. 

(This is not important but we are doing this just to make an extra commit and see if Circle CI passes.)

**Review by rt4914 — CHANGES_REQUESTED** [review]
I have added one comment to check if circle CI passes this time.  Also one more thing, merge your code with the latest develop.

**Comment by Sarthak2601** [comment]
Hi @aggarwalpulkit596, I had a similar issue in one of my previous PRs. In my case, closing the PR and creating a new one worked. Maybe it can work for you too. :)

**Comment by aggarwalpulkit596** [comment]
@Sarthak2601 instead of this we should create a better emulator which is more responsive for UI test

**Comment by rt4914** [comment]
> Hi @aggarwalpulkit596, I had a similar issue in one of my previous PRs. In my case, closing the PR and creating a new one worked. Maybe it can work for you too. :)

Actually this issue is coming from past 1 week only and it is temporary issue it will get resolved soon. But till then we are just this so that no one gets blocked.

**Review by rt4914 — APPROVED** [review]
LGTM

**Pull request merged**